out_stackdriver: add static labels defined in configuration

franciscovalentecastro opened the pull request:
This change mirrors the [out_loki](https://docs.fluentbit.io/manual/pipeline/outputs/loki#using-the-label_keys-property) exporter functionality to set static `labels` during configuration. This as a list of comma separate `key=value` pairs, e.g. `label_a=value_a,label_b=value_b`. The `record_accesor` wasn't implemented since the functionality can be recreated using `label_keys`.

The expected behavior is that elements set with `labels` and `label_keys` are combined to set the LogEntry Labels in the LogExplorer. When duplicated keys appear, we would give preference to values set in `labels` over `label_keys`.

This PR is one approach to add the feature in #4912 .

### Configuration

This configuration implements the functionality by setting `labels` using `key=value` pairs and using a `modify` + `nest` to set a label in the payload to show the output combines the result.

```
[INPUT]
    Buffer_Chunk_Size 512k
    Buffer_Max_Size   5M
    Key               message
    Mem_Buf_Limit     10M
    Name              tail
    Path              /var/log/messages,/var/log/syslog
    Read_from_Head    True
    Rotate_Wait       30
    Skip_Long_Lines   On
    Tag               default_pipeline.syslog

[FILTER]
    Add   labelA valueA
    Match *
    Name  modify

[FILTER]
    Match      *
    Name       nest
    Nest_under logging.googleapis.com/labels
    Operation  nest
    Wildcard   labelA*

[OUTPUT]
    Match_Regex                   ^(default_pipeline\.syslog)$
    Name                          stackdriver
    Retry_Limit                   3
    net.connect_timeout_log_error False
    resource                      gce_instance
    stackdriver_agent             Google-Cloud-Ops-Agent-Logging/2.11.0 (BuildDistro=buster;Platform=linux;ShortName=debian;ShortVersion=10.11)
    tls                           On
    tls.verify                    Off
    workers                       8
    labels                        compute.googleapis.com/resource_name=testing_vm_syslog,labelQ=valueQ
```

This is one output `LogEntry` :

![Screen Shot 2022-03-28 at 13 00 41](https://user-images.githubusercontent.com/1435136/160468154-23e1f5e9-45f4-43d0-99f7-2737bc2b65b9.png)



### Output Log and Valgrind result
The result of the following `valgrind` execution is shown below :
```
valgrind --leak-check=yes fluent-bit/build/bin/fluent-bit   --config ./fluent_bit_main.conf
```

This is the output log of the execution : 

```
==18343== Memcheck, a memory error detector
==18343== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==18343== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==18343== Command: fluent-bit/build/bin/fluent-bit --config ./fluent_bit_main.conf
==18343== 
Fluent Bit v1.9.3
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/04/11 16:39:41] [ info] [fluent bit] version=1.9.3, commit=3d5d754ffa, pid=18343
[2022/04/11 16:39:41] [ info] [storage] version=1.1.6, type=memory-only, sync=normal, checksum=disabled, max_chunks_up=128
[2022/04/11 16:39:41] [ info] [cmetrics] version=0.3.0
[2022/04/11 16:39:44] [ info] [output:stackdriver:stackdriver.0] metadata_server set to http://metadata.google.internal
[2022/04/11 16:39:44] [ warn] [output:stackdriver:stackdriver.0] client_email is not defined, using a default one
[2022/04/11 16:39:44] [ warn] [output:stackdriver:stackdriver.0] private_key is not defined, fetching it from metadata server
[2022/04/11 16:39:46] [ info] [output:stackdriver:stackdriver.0] worker #0 started
[2022/04/11 16:39:46] [ info] [output:stackdriver:stackdriver.0] worker #1 started
[2022/04/11 16:39:46] [ info] [output:stackdriver:stackdriver.0] worker #2 started
[2022/04/11 16:39:46] [ info] [output:stackdriver:stackdriver.0] worker #3 started
[2022/04/11 16:39:46] [ info] [output:stackdriver:stackdriver.0] worker #4 started
[2022/04/11 16:39:46] [ info] [output:stackdriver:stackdriver.0] worker #5 started
[2022/04/11 16:39:46] [ info] [output:stackdriver:stackdriver.0] worker #6 started
[2022/04/11 16:39:46] [ info] [output:stackdriver:stackdriver.0] worker #7 started
[2022/04/11 16:39:46] [ info] [sp] stream processor started
[2022/04/11 16:39:47] [ info] [input:tail:tail.0] inotify_fs_add(): inode=410658 watch_fd=1 name=/var/log/messages
[2022/04/11 16:39:47] [ info] [input:tail:tail.0] inotify_fs_add(): inode=407396 watch_fd=2 name=/var/log/syslog
^C[2022/04/11 16:40:13] [engine] caught signal (SIGINT)
[2022/04/11 16:40:13] [ info] [input] pausing tail.0
[2022/04/11 16:40:13] [ warn] [engine] service will shutdown in max 5 seconds
[2022/04/11 16:40:13] [ info] [engine] service has stopped (0 pending tasks)
[2022/04/11 16:40:13] [ info] [input:tail:tail.0] inotify_fs_remove(): inode=410658 watch_fd=1
[2022/04/11 16:40:13] [ info] [input:tail:tail.0] inotify_fs_remove(): inode=407396 watch_fd=2
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #0 stopping...
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #0 stopped
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #1 stopping...
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #1 stopped
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #2 stopping...
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #2 stopped
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #3 stopping...
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #3 stopped
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #4 stopping...
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #4 stopped
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #5 stopping...
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #5 stopped
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #6 stopping...
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #6 stopped
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #7 stopping...
[2022/04/11 16:40:13] [ info] [output:stackdriver:stackdriver.0] thread worker #7 stopped
==18343== 
==18343== HEAP SUMMARY:
==18343==     in use at exit: 0 bytes in 0 blocks
==18343==   total heap usage: 69,777 allocs, 69,777 frees, 116,382,573 bytes allocated
==18343== 
==18343== All heap blocks were freed -- no leaks are possible
==18343== 
==18343== For counts of detected and suppressed errors, rerun with: -v
==18343== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

Signed-off-by: Francisco Valente <fcovalente@google.com>

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change
<!-- Invoke Fluent Bit and Valgrind as: $ valgrind ./bin/fluent-bit <args> -->
- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [x] Documentation required for this feature
Documentation PR #https://github.com/fluent/fluent-bit-docs/pull/768.

<!--  Doc PR (not required but highly recommended) -->

**Backporting**
<!--
PRs targeting the default master branch will go into the next major release usually.
If this PR should be backported to the current or earlier releases then please submit a PR for that particular branch.
-->
- [x] Backport to latest stable release.

The PR for the backport into `1.8` branch #5095.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

edsiper commented:
CI is failing:

https://github.com/fluent/fluent-bit/runs/5833671590?check_suite_focus=true#step:5:3092

also, can you rebase on top of master ?

franciscovalentecastro commented:
I rebased on top of `master` and added commit https://github.com/fluent/fluent-bit/pull/5176/commits/a4708cb2aa729bdca0c92f4c1390c0d5a2e24ec4 that addressed one of the issues of the `PR`. CI is still failing, I'm exploring the possible issues.

franciscovalentecastro commented:
Hello @edsiper. I updated the PR fixing the errors that were causing the CI failure. Some notes : 

* Verified that locally that all unit tests pass now.
* Rebased last changes from master.
* Updated PR comments and valgrind tests.
